libs/motor_control/esc/esc.c

```diff
@@ -0,0 +1,222 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, version 3.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/**
+ * @file   esc.c
+ * @brief  Zephyr Electronic Speed Controller ESC driver for BLDC motors
+ * @author Eurus-project
+ */
+
+/******************************** Includes ************************************/
+#include "esc.h"
+#include <stdint.h>
+#include <stdlib.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/devicetree/pinctrl.h>
+#include <zephyr/drivers/pwm.h>
+#include <zephyr/kernel.h>
+
+/****************************** Static Functions ******************************/
+static status_t prvESC_ChannelCheck(const struct device *pwm_dev,
+                                    uint32_t pwm_channel);
+
+/******************************** Functions ***********************************/
+
+/* TODO: Retreive some info from device tree about pwm channels
+ */
+status_t ESC_Init(const struct device *pwm_dev, uint32_t pwm_channel,
+                  esc_protocol_t protocol, esc_t *esc_out) {
+    static esc_number_t esc_instance_num = 0;
+
+    if (!device_is_ready(pwm_dev))
+        return -1;
+
+    switch (protocol) {
+    case ESC_PWM:
+        esc_out->period = T_PWM_US;
+        esc_out->dc_min = DC_MIN_PWM_US;
+        esc_out->dc_max = DC_MAX_PWM_US;
+        esc_out->protocol = ESC_PWM;
+        break;
+
+    case ESC_ONESHOT_125:
+        esc_out->period = T_ONESHOT_125;
+        esc_out->dc_min = DC_MIN_ONESHOT_125_US;
+        esc_out->dc_max = DC_MAX_ONESHOT_125_US;
+        esc_out->protocol = ESC_ONESHOT_125;
+        break;
+
+    case ESC_ONESHOT_42:
+        esc_out->period = T_ONESHOT_42;
+        esc_out->dc_min = DC_MIN_ONESHOT_42_US;
+        esc_out->dc_max = DC_MAX_ONESHOT_42_US;
+        esc_out->protocol = ESC_ONESHOT_42;
+        break;
+
+    case ESC_MULTISHOT:
+        esc_out->period = T_MULTISHOT;
+        esc_out->dc_min = DC_MIN_MULTISHOT_US;
+        esc_out->dc_max = DC_MAX_MULTISHOT_US;
+        esc_out->protocol = ESC_MULTISHOT;
+        break;
+
+    default:
+        esc_out->flag = ESC_UNINITIALIZED;
+        return -1;
+    }
+    esc_out->instance.pwm_dev = pwm_dev;
+    esc_out->instance.pwm_channel = pwm_channel;
+    esc_out->instance.instance_num = esc_instance_num;
+    esc_out->flag = ESC_INITIALIZED;
+    esc_instance_num++;
+
+    return 0;
+}
+
+status_t ESC_SetSpeed(esc_t *esc, uint8_t speed) {
+    int ret;
+    uint32_t dc_speed;
+
+    if (esc->flag != ESC_INITIALIZED) {
+        printk("[ESC]: Error: Uninitialized ESC cannot be started!");
```

Comment:
As @nikolaptr has already mentioned, we do not need these, but in cases where you do want to print errors in a library, please use Zephyrs logging subsystem.